File: sort/hsort.cpp
#include <iostream>
#include <ctime>
#include <cstdlib>
#include <iomanip>
#include <cstring>
#include <algorithm>

using namespace std;
const int N = 10;

void adjust(int a[], int len, int index){
    int left = 2*index + 1;
    int right = 2*index + 2;
    int maxIdx = index;
    if(left<len && a[left] > a[maxIdx]) maxIdx = left;
    if(right<len && a[right] > a[maxIdx])maxIdx = right;
    if(maxIdx != index){
        swap(a[maxIdx],a[index]);
        adjust(a, len, maxIdx);
    }
}

void hsort(int a[], int size){
    for(int i=size/2 - 1; i>=0; i--){
        adjust(a,size,i);
    }

    for(int i=size-1; i>0; i--){
        swap(a[0],a[i]);
        adjust(a,i,0);
    }
}


int main(){
    int a[N];
    srand(time(NULL));
    for(int i=0; i<N; i++){
        a[i]=rand()%100;
        cout << setw(3) <<a[i];
    }
    cout << endl;

    hsort(a,N);

    for(int i=0; i<N; i++){
        cout << setw(3) << a[i];
    }
    cout << endl;

    return 0;
}

File: sort/hsort.py
#!/usr/bin/env python
__author__ = 'indigo'

import numpy as np

class HeapSort(object):
    def __init__(self,n):
        self.L = np.random.randint(0,99,n)

    def swap(self,L,m,n):
        L[m]^=L[n]
        L[n]^=L[m]
        L[m]^=L[n]

    def adjust(self,L,length,index):
        left = 2*index+1
        right = 2*index+2
        maxIdx = index

        if left<length and L[left]>L[maxIdx]:
            maxIdx = left
        if right<length and L[right]>L[maxIdx]:
            maxIdx = right

        if maxIdx != index:
            self.swap(L,index,maxIdx)
            self.adjust(L,length,maxIdx)

    def heapsort(self):
        L = self.L
        length = len(L)
        for i in range(length/2-1,-1,-1):
            self.adjust(L,length,i)

        for i in range(length-1,0,-1):
            self.swap(L,i,0)
            self.adjust(L,i,0)

    def show(self):
        print self.L

if __name__ == "__main__":
    t = HeapSort(10)
    t.show()
    t.heapsort()
    t.show()
